Add patching pass to annotate discard to demote requirements.

perlfu opened the pull request:
While the Vulkan standard indicates that implicit derivatives are
undefined after a discard operation, many shaders attempt early
discard in ways which leads to undefined behaviour.
Attempt to detect these cases by searching forward in program
order from blocks which conditionally reach discards looking for
image operations which require implicit derivatives.
If any such path exists from a discard to implicit derivative use,
then annotate the fragment shader with
"amdgpu-transform-discard-to-demote".
This will trigger the backend AMDGPUConditionalDiscard pass,
when enabled, to attempt to replace discards with demotes.
(Demotes have defined behaviour for implicit derivatives.)

dstutt commented on lgc/patch/PatchImageDerivatives.cpp:
```diff
@@ -0,0 +1,201 @@
+
+    while (!worklist.empty()) {
+      BasicBlock *potentialRoot = worklist.pop_back_val();
+      if (visitedBlocks.count(potentialRoot) > 0)
```
I don't think you need the > 0 here do you?

dstutt commented on lgc/patch/PatchImageDerivatives.cpp:
```diff
@@ -0,0 +1,201 @@
+  }
+  assert(worklist.empty());
+
+  // Breath first search from roots looking for any block containing derivatives.
```
Breath -> Breadth

dstutt commented on lgc/patch/PatchImageDerivatives.cpp:
```diff
@@ -0,0 +1,201 @@
+      continue;
+
+    visitedBlocks.insert(testBlock);
+    if (derivativeBlocks.count(testBlock) > 0) {
```
> 0 not required?

dstutt commented on llpc/test/shaderdb/general/DiscardToDemoteTransformations.frag:
```diff
@@ -1,7 +1,8 @@
-// Check that amdllpc does not enable discard-to-demote transforms by default and that
-// they can be enabled on demand. This should affect the generated code and cache hash.
+// Check that amdllpc enables discard-to-demote transforms by automatically and that they
```
by automatically -> automatically

dstutt reviewed (dismissed):
This LGTM
(See minor nits)

I'd guess that the cost of this pass shouldn't be high - and I think I understand the algorithm you're using (and afaict it should always complete)?

dstutt commented:
retest this please

amdvlk-admin commented:
### Test summary for commit 86b9da821a0e20a5b1eecac454b15edd0507947b
<details><summary>CTS tests (Failed: 0/289625)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx9</summary><ul>

<li>Passed: 32868/69268 (47.5%)</li>
<li>Failed: 0/69268 (0.0%)</li>
<li>Not Supported: 36400/69268 (52.5%)</li><li>Warnings: 0/69268 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi21</summary><ul>

<li>Passed: 38491/73498 (52.4%)</li>
<li>Failed: 0/73498 (0.0%)</li>
<li>Not Supported: 35007/73498 (47.6%)</li><li>Warnings: 0/73498 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 34348/73361 (46.8%)</li>
<li>Failed: 0/73361 (0.0%)</li>
<li>Not Supported: 39013/73361 (53.2%)</li><li>Warnings: 0/73361 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 38445/73498 (52.3%)</li>
<li>Failed: 0/73498 (0.0%)</li>
<li>Not Supported: 35053/73498 (47.7%)</li><li>Warnings: 0/73498 (0.0%)</li>
</ul></details>


</ul></details>

amdvlk-admin commented:
### Test summary for commit c8999253d7aa2f489f37274d32b466f2321e7afe
<details><summary>CTS tests (Failed: 0/289625)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx9</summary><ul>

<li>Passed: 32868/69268 (47.5%)</li>
<li>Failed: 0/69268 (0.0%)</li>
<li>Not Supported: 36400/69268 (52.5%)</li><li>Warnings: 0/69268 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi21</summary><ul>

<li>Passed: 38491/73498 (52.4%)</li>
<li>Failed: 0/73498 (0.0%)</li>
<li>Not Supported: 35007/73498 (47.6%)</li><li>Warnings: 0/73498 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 34348/73361 (46.8%)</li>
<li>Failed: 0/73361 (0.0%)</li>
<li>Not Supported: 39013/73361 (53.2%)</li><li>Warnings: 0/73361 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35388/73498 (48.1%)</li>
<li>Failed: 0/73498 (0.0%)</li>
<li>Not Supported: 38110/73498 (51.9%)</li><li>Warnings: 0/73498 (0.0%)</li>
</ul></details>


</ul></details>

perlfu commented:
> This LGTM (See minor nits)
> 
> I'd guess that the cost of this pass shouldn't be high - and I think I understand the algorithm you're using (and afaict it should always complete)?

Thanks, I think I have addressed your nits.

I have tried to keep this as efficient as possible.
Worst case it visits every block in the shader, and will always complete.
It's unfortunate that proving a shader valid requires more processing than detecting it is invalid.
However, in a valid shader the discards do tend to appear later so this offsets the cost.

Testing on our regular catalogue of shaders this visited only 4 blocks on average, with the worst case being 35 blocks visited.
So in practice the performance impact should be very low.

I am going to do some more testing on this as I feel like it is potentially labelling demotes as required too frequently.

perlfu commented:
Having looked through several examples I am now fairly confident this is correct.
We just seem to have a lot of shaders with undefined behaviour in our database.

dstutt reviewed (approved):
Thanks. LGTM too.
